packages/next/src/client/components/react-dev-overlay/README.md

```diff
@@ -1,10 +1,71 @@
-TODO
+# Dev Overlay
 
-### Folder Structure
+## Project Structure
 
-- `ui` - This should be isolated from the rest of the codebase to make it easier to be excluded from compiliation and be bundled.
+- `app/` - Main entrypoint and related files for App Router dev overlay.
+- `font/` - Includes Geist fonts that will be served to the client via the internal `/__nextjs_font/...` route.
+- `pages/` - Main entrypoint and related files for Pages Router dev overlay.
+- `server/` - Source files related to the dev overlay that are to run on the server and not sent to the client.
+- `ui/` - Main source files that build the UI of the dev overlay. It is recommended to keep this directory isolated from the rest of the codebase to make it easier to be excluded from compiliation and be bundled in the future.
+- `utils/` - Utils that can also be used outside of this project unlike the ones inside the `ui/` directory.
+
+## Development
+
+The dev overlay leverages [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM) to insert the dev overlay into the client without affecting the user's application through encapsulation from CSS.
+
+### Local Development
+
+Next.js uses [Storybook](https://storybook.js.org) to develop UI components in the dev overlay.
+
+To run the dev overlay locally, you can run the following command:
+
+```bash
+pnpm storybook
+```
+
+This will start the Storybook server at `http://localhost:6006`.
 
 ### Styling
 
-- The entrypoint needs to use `css` util that minifies styles.
-  - Entrypoint is which are using `<style>` tag to inject styles.
+Next.js direcly injects CSS into the DOM via `<style>` tag. The styles will not affect the user's application as the [styles are encapsulated](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#encapsulation_from_css) from the rest of the application.
+
+> [!TIP]
+> While Shadow DOM provides style encapsulation, the root element (i.e., `<nextjs-portal>`) can still inherit styles from parent elements like `<body>` or `<html>`. Direct styling on these parent elements (e.g., `body { contain: layout; }`) will affect the dev overlay.
+
+Write CSS in template literals alongside your components. It is recommended to use the class names that are unique within the dev overlay to avoid conflicts. You can use `data-nextjs-` data attributes to target the elements in the dev overlay.
```

Comment:
Something that would be nice for us to standardize on is data attributes vs classNames for styling. Sometimes we do one or the other, sometimes we do both.